Index: socat-1.6.0.0/sysincludes.h
===================================================================
--- socat-1.6.0.0.orig/sysincludes.h	2007-03-06 13:04:58.000000000 -0800
+++ socat-1.6.0.0/sysincludes.h	2007-06-08 17:57:52.000000000 -0700
@@ -141,9 +141,6 @@
 #if HAVE_LINUX_FS_H
 #include <linux/fs.h>		/* somewhere required for ext2_fs.h */
 #endif
-#if HAVE_LINUX_EXT2_FS_H
-#include <linux/ext2_fs.h>	/* Linux ext2 filesystem definitions */
-#endif
 #if WITH_READLINE
 #  if HAVE_READLINE_READLINE_H
 #include <readline/readline.h>
