diff -ruN net-snmp-5.2.1.2.orig/snmplib/snmpUDPIPv6Domain.c net-snmp-5.2.1.2/snmplib/snmpUDPIPv6Domain.c
--- net-snmp-5.2.1.2.orig/snmplib/snmpUDPIPv6Domain.c	2004-12-24 06:42:23.000000000 -0800
+++ net-snmp-5.2.1.2/snmplib/snmpUDPIPv6Domain.c	2009-03-14 18:20:50.000000000 -0700
@@ -1125,8 +1125,8 @@
                      0, NI_NUMERICHOST)) {
                     config_perror("getnameinfo failed");
                 }
-                memmove(ai->ai_addr, &net, sizeof(struct sockaddr_in6));
-                inet_make_mask_addr(AF_INET6, &mask.sin6_addr, 127);
+                memmove(&net, ai->ai_addr, sizeof(struct sockaddr_in6));
+                inet_make_mask_addr(PF_INET6, &mask.sin6_addr, 128);
 
                 e = (com2Sec6Entry *) malloc(sizeof(com2Sec6Entry));
                 if (e == NULL) {
@@ -1244,7 +1244,7 @@
 
     for (c = com2Sec6List; c != NULL; c = c->next) {
         DEBUGMSGTL(("netsnmp_udp6_getSecName",
-                    "compare <\"%s\", 0x%032/0x%032x>", c->community,
+                    "compare <\"%s\", 0x%032x/0x%032x>", c->community,
                     c->network, c->mask));
 
         if ((community_len == (int)strlen(c->community)) &&
