Index: httpd-2.2.14/srclib/apr/configure
===================================================================
--- httpd-2.2.14.orig/srclib/apr/configure
+++ httpd-2.2.14/srclib/apr/configure
@@ -27075,7 +27075,7 @@ $as_echo_n "checking for /dev/zero... " 
 if test "${ac_cv_file__dev_zero+set}" = set; then :
   $as_echo_n "(cached) " >&6
 else
-  test "$cross_compiling" = yes &&
+  test "$cross_compiling" = no &&
   as_fn_error "cannot check for file existence when cross compiling" "$LINENO" 5
 if test -r "/dev/zero"; then
   ac_cv_file__dev_zero=yes
@@ -30285,7 +30285,7 @@ if test "${ac_cv_sizeof_struct_iovec+set
   $as_echo_n "(cached) " >&6
 else
   if test "$cross_compiling" = yes; then :
-  ac_cv_sizeof_struct_iovec=0
+  ac_cv_sizeof_struct_iovec="(sizeof(struct iovec))"
 else
   cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
@@ -33337,7 +33337,7 @@ $as_echo_n "checking whether TCP_NODELAY
 if test "${apr_cv_tcp_nodelay_with_cork+set}" = set; then :
   $as_echo_n "(cached) " >&6
 else
-  if test "$cross_compiling" = yes; then :
+  if test "$cross_compiling" = no; then :
   { { $as_echo "$as_me:${as_lineno-$LINENO}: error: in \`$ac_pwd':" >&5
 $as_echo "$as_me: error: in \`$ac_pwd':" >&2;}
 as_fn_error "cannot run test program while cross compiling
Index: httpd-2.2.14/configure
===================================================================
--- httpd-2.2.14.orig/configure
+++ httpd-2.2.14/configure
@@ -7005,7 +7005,7 @@ if test "${ap_cv_void_ptr_lt_long+set}" 
   $as_echo_n "(cached) " >&6
 else
   if test "$cross_compiling" = yes; then :
-  ap_cv_void_ptr_lt_long=yes
+  ap_cv_void_ptr_lt_long=no
 else
   cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
