--- /u-boot-1.1.2/Makefile	2007-07-17 08:14:46.000000000 -0500
+++ u-boot-1.1.2/Makefile	2007-07-18 05:17:34.000000000 -0500
@@ -158,8 +158,9 @@
 		$(OBJDUMP) -d $< > $@
 
 u-boot:		depend $(SUBDIRS) $(OBJS) $(LIBS) $(LDSCRIPT)
+		FLAGS=`echo "$(LDFLAGS)"|sed -r 's/\-Wl,(.*),/\1 /'`;\
 		UNDEF_SYM=`$(OBJDUMP) -x $(LIBS) |sed  -n -e 's/.*\(__u_boot_cmd_.*\)/-u\1/p'|sort|uniq`;\
-		$(LD) $(LDFLAGS) $$UNDEF_SYM $(OBJS) \
+		$(LD) $$FLAGS $$UNDEF_SYM $(OBJS) \
 			--start-group $(LIBS) $(PLATFORM_LIBS) --end-group \
 			-Map u-boot.map -o u-boot
 
