Index: postgresql-8.1.4/src/interfaces/ecpg/compatlib/Makefile
===================================================================
--- postgresql-8.1.4.orig/src/interfaces/ecpg/compatlib/Makefile
+++ postgresql-8.1.4/src/interfaces/ecpg/compatlib/Makefile
@@ -20,6 +20,7 @@ DLTYPE= library
 override CPPFLAGS := -I$(top_srcdir)/src/interfaces/ecpg/include -I$(libpq_srcdir) \
 	-I$(top_srcdir)/src/include/utils $(CPPFLAGS)
 override CFLAGS += $(PTHREAD_CFLAGS)
+override CFLAGS := -I../include $(CFLAGS)
 SHLIB_LINK = -L../ecpglib -lecpg -L../pgtypeslib -lpgtypes $(libpq) \
 	$(filter -lintl -lssl -lcrypto -lkrb5 -lcrypt -lm, $(LIBS)) $(PTHREAD_LIBS)
 
