Index: linux-2.6.20/arch/powerpc/kernel/setup-common.c
===================================================================
--- linux-2.6.20.orig/arch/powerpc/kernel/setup-common.c	2008-11-05 15:49:25.000000000 -0800
+++ linux-2.6.20/arch/powerpc/kernel/setup-common.c	2008-11-05 15:48:55.000000000 -0800
@@ -67,6 +67,8 @@
 #define DBG(fmt...)
 #endif
 
+extern unsigned long wdt_base;
+
 /* The main machine-dep calls structure
  */
 struct machdep_calls ppc_md;
@@ -116,6 +118,7 @@
 
 void machine_power_off(void)
 {
+	int i;
 	machine_shutdown();
 	if (ppc_md.power_off)
 		ppc_md.power_off();
@@ -124,7 +127,15 @@
 #endif
 	printk(KERN_EMERG "System Halted, OK to turn off power\n");
 	local_irq_disable();
-	while (1) ;
+	while (1) {
+	//Added for BTS Bug-51589, Shell commands like shutdown or halt should NOT reboot the PM   
+	#ifdef CONFIG_PPC_83xx
+		out_be16((__be16 *)(wdt_base + 0xe), 0x556c);
+		out_be16((__be16 *)(wdt_base + 0xe), 0xaa39);
+	#endif
+	for(i =0; i < 1000; i++ )
+		;
+	}
 }
 /* Used by the G5 thermal driver */
 EXPORT_SYMBOL_GPL(machine_power_off);
