Index: Linux-PAM-1.0.2/modules/pam_securetty/pam_securetty.c
===================================================================
--- Linux-PAM-1.0.2.orig/modules/pam_securetty/pam_securetty.c
+++ Linux-PAM-1.0.2/modules/pam_securetty/pam_securetty.c
@@ -87,7 +87,7 @@ securetty_perform_check (pam_handle_t *p
 
     user_pwd = pam_modutil_getpwnam(pamh, username);
     if (user_pwd == NULL) {
-	return PAM_USER_UNKNOWN;
+	return PAM_IGNORE;
     } else if (user_pwd->pw_uid != 0) { /* If the user is not root,
 					   securetty's does not apply
 					   to them */
Index: Linux-PAM-1.0.2/libpam_misc/misc_conv.c
===================================================================
--- Linux-PAM-1.0.2.orig/libpam_misc/misc_conv.c
+++ Linux-PAM-1.0.2/libpam_misc/misc_conv.c
@@ -189,6 +189,9 @@ static int read_string(int echo, const c
 			    nc = rv;
 			break;
 		    }
+		    if (line[nc] == '\r') {
+			line[nc] = '\n';
+		    }
 		}
 	    if (have_term) {
 		(void) tcsetattr(STDIN_FILENO, TCSADRAIN, &term_before);
