Index: Linux-PAM-1.0.2/Makefile.in
===================================================================
--- Linux-PAM-1.0.2.orig/Makefile.in
+++ Linux-PAM-1.0.2/Makefile.in
@@ -71,7 +71,7 @@ RECURSIVE_CLEAN_TARGETS = mostlyclean-re
   distclean-recursive maintainer-clean-recursive
 ETAGS = etags
 CTAGS = ctags
-DIST_SUBDIRS = libpam tests libpamc libpam_misc modules po conf doc \
+DIST_SUBDIRS = libpam tests libpamc libpam_misc modules po conf \
 	examples xtests
 DISTFILES = $(DIST_COMMON) $(DIST_SOURCES) $(TEXINFOS) $(EXTRA_DIST)
 distdir = $(PACKAGE)-$(VERSION)
@@ -244,8 +244,8 @@ target_alias = @target_alias@
 top_builddir = @top_builddir@
 top_srcdir = @top_srcdir@
 AUTOMAKE_OPTIONS = 1.9 gnu dist-bzip2 check-news
-@STATIC_MODULES_FALSE@SUBDIRS = libpam tests libpamc libpam_misc modules po conf doc examples xtests
-@STATIC_MODULES_TRUE@SUBDIRS = modules libpam libpamc libpam_misc tests po conf doc examples xtests
+@STATIC_MODULES_FALSE@SUBDIRS = libpam tests libpamc libpam_misc modules po conf examples xtests
+@STATIC_MODULES_TRUE@SUBDIRS = modules libpam libpamc libpam_misc tests po conf  examples xtests
 CLEANFILES = *~
 M4_FILES = m4/codeset.m4 m4/gettext.m4 m4/glibc21.m4 m4/glibc2.m4 \
 	m4/iconv.m4 m4/intdiv0.m4 m4/intmax.m4 m4/inttypes_h.m4 \
@@ -718,19 +718,7 @@ uninstall-am:
 	tags tags-recursive uninstall uninstall-am
 
 
-release: dist releasedocs
-
-release-docs: releasedocs
-
-releasedocs:
-	rm -rf Linux-PAM-$(VERSION)
-	mkdir -p Linux-PAM-$(VERSION)/doc
-	make -C doc releasedocs
-	tar zfc Linux-PAM-$(VERSION)-docs.tar.gz \
-		Linux-PAM-$(VERSION)/doc
-	tar jfc Linux-PAM-$(VERSION)-docs.tar.bz2 \
-		Linux-PAM-$(VERSION)/doc
-	rm -rf Linux-PAM-$(VERSION)
+release: dist 
 
 xtests:
 	make -C xtests xtests
