Index: u-boot-1.2.0/cpu/ppc4xx/4xx_enet.c
===================================================================
--- u-boot-1.2.0.orig/cpu/ppc4xx/4xx_enet.c
+++ u-boot-1.2.0/cpu/ppc4xx/4xx_enet.c
@@ -566,6 +567,20 @@ static int ppc_4xx_eth_init (struct eth_
 		miiphy_write (dev->name, reg, 0x09, 0x0e00);
 		miiphy_write (dev->name, reg, 0x04, 0x01e1);
 #endif
+
+#if defined(CONFIG_M88E1121_PHY)
+		miiphy_write (dev->name, reg, 22, 2); 		//set page 2
+		miiphy_write (dev->name, reg, 21, 0x1070);
+
+		miiphy_write (dev->name, reg, 22, 0);       //set page 0
+		miiphy_write (dev->name, reg, 16, 0x6060);
+
+		miiphy_write (dev->name, reg, 22, 3);      	//set page 3
+		miiphy_write (dev->name, reg, 16, 0x101e);
+
+		miiphy_write (dev->name, reg, 0x09, 0x0e00);
+		miiphy_write (dev->name, reg, 0x04, 0x01e1);
+#endif
 		miiphy_reset (dev->name, reg);
 
 #if defined(CONFIG_440GX) || \
