Index: u-boot-1.3.1/Makefile
===================================================================
--- u-boot-1.3.1.orig/Makefile	2008-01-31 17:25:23.000000000 -0800
+++ u-boot-1.3.1/Makefile	2008-01-31 17:32:27.000000000 -0800
@@ -292,8 +292,9 @@
 		$(OBJDUMP) -d $< > $@
 
 $(obj)u-boot:		depend version $(SUBDIRS) $(OBJS) $(LIBS) $(LDSCRIPT)
+		FLAGS=`echo "$(LDFLAGS)"|sed -r 's/\-Wl,(.*),/\1 /'`;\
 		UNDEF_SYM=`$(OBJDUMP) -x $(LIBS) |sed  -n -e 's/.*\(__u_boot_cmd_.*\)/-u\1/p'|sort|uniq`;\
-		cd $(LNDIR) && $(LD) $(LDFLAGS) $$UNDEF_SYM $(__OBJS) \
+		cd $(LNDIR) && $(LD) $$FLAGS $$UNDEF_SYM $(__OBJS) \
 			--start-group $(__LIBS) --end-group $(PLATFORM_LIBS) \
 			-Map u-boot.map -o u-boot
 
