Index: glibc-2.3.5/linuxthreads/sysdeps/pthread/bits/typesizes.h
===================================================================
--- glibc-2.3.5.orig/linuxthreads/sysdeps/pthread/bits/typesizes.h
+++ glibc-2.3.5/linuxthreads/sysdeps/pthread/bits/typesizes.h
@@ -60,7 +60,7 @@
 #define __SSIZE_T_TYPE		__SWORD_TYPE
 
 /* Number of descriptors that can fit in an `fd_set'.  */
-#define	__FD_SETSIZE		1024
+#define	__FD_SETSIZE		8192
 
 
 #endif /* bits/typesizes.h */
Index: glibc-2.3.5/sysdeps/generic/bits/typesizes.h
===================================================================
--- glibc-2.3.5.orig/sysdeps/generic/bits/typesizes.h
+++ glibc-2.3.5/sysdeps/generic/bits/typesizes.h
@@ -60,7 +60,7 @@
 #define __SSIZE_T_TYPE		__SWORD_TYPE
 
 /* Number of descriptors that can fit in an `fd_set'.  */
-#define	__FD_SETSIZE		1024
+#define	__FD_SETSIZE		8192
 
 
 #endif /* bits/typesizes.h */
